File: integrations/dotnet/VoiceMdClient.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace VoiceMd;

public sealed record VoiceSelector(
    string? Profile = null,
    string? Audience = null,
    string? Surface = null,
    string? Tone = null,
    string? Format = null,
    bool? Compact = null,
    int? MaxChars = null);

public sealed record LintIssue(
    [property: JsonPropertyName("rule_id")] string RuleId,
    string Severity,
    string Message,
    string? Evidence);

public sealed record LintResult(bool Ok, IReadOnlyList<LintIssue> Issues);

public sealed class VoiceMdClient(HttpClient httpClient)
{
    public async Task<string> CompileAsync(VoiceSelector? selector = null, CancellationToken cancellationToken = default)
    {
        selector ??= new VoiceSelector();
        var query = new List<string>();
        Add(query, "profile", selector.Profile);
        Add(query, "audience", selector.Audience);
        Add(query, "surface", selector.Surface);
        Add(query, "tone", selector.Tone);
        Add(query, "format", selector.Format);
        Add(query, "compact", selector.Compact?.ToString().ToLowerInvariant());
        Add(query, "max_chars", selector.MaxChars?.ToString());

        var path = "/v1/voice/prompt" + (query.Count == 0 ? "" : "?" + string.Join("&", query));
        using var response = await httpClient.GetAsync(path, cancellationToken);
        response.EnsureSuccessStatusCode();
        var payload = await response.Content.ReadFromJsonAsync<CompileResponse>(cancellationToken: cancellationToken)
            ?? throw new InvalidOperationException("VoiceMD returned an empty response.");
        return payload.Prompt;
    }

    public async Task<LintResult> LintAsync(
        string text,
        VoiceSelector? selector = null,
        CancellationToken cancellationToken = default)
    {
        selector ??= new VoiceSelector();
        using var response = await httpClient.PostAsJsonAsync(
            "/v1/voice/lint",
            new
            {
                text,
                profile = selector.Profile,
                audience = selector.Audience,
                surface = selector.Surface,
                tone = selector.Tone
            },
            cancellationToken);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<LintResult>(cancellationToken: cancellationToken)
            ?? throw new InvalidOperationException("VoiceMD returned an empty response.");
    }

    private static void Add(List<string> query, string name, string? value)
    {
        if (!string.IsNullOrWhiteSpace(value))
            query.Add($"{Uri.EscapeDataString(name)}={Uri.EscapeDataString(value)}");
    }

    private sealed record CompileResponse(string Prompt);
}

File: integrations/typescript/voicemd-client.ts
export type VoiceSelector = {
  profile?: string;
  audience?: string;
  surface?: string;
  tone?: string;
  format?: "prompt" | "json" | "ascii" | "nemotron" | "nemotron-ascii";
  compact?: boolean;
  maxChars?: number;
};

export type LintIssue = {
  rule_id: string;
  severity: "info" | "warning" | "error" | string;
  message: string;
  evidence?: string | null;
};

export class VoiceMdClient {
  constructor(private readonly baseUrl = "http://127.0.0.1:8765") {}

  async compile(selector: VoiceSelector = {}): Promise<string> {
    const url = new URL("/v1/voice/prompt", this.baseUrl);
    if (selector.profile) url.searchParams.set("profile", selector.profile);
    if (selector.audience) url.searchParams.set("audience", selector.audience);
    if (selector.surface) url.searchParams.set("surface", selector.surface);
    if (selector.tone) url.searchParams.set("tone", selector.tone);
    if (selector.format) url.searchParams.set("format", selector.format);
    if (selector.compact !== undefined) url.searchParams.set("compact", String(selector.compact));
    if (selector.maxChars !== undefined) url.searchParams.set("max_chars", String(selector.maxChars));

    const response = await fetch(url);
    if (!response.ok) throw new Error(`VoiceMD compile failed: ${response.status} ${await response.text()}`);
    const payload = (await response.json()) as { prompt: string };
    return payload.prompt;
  }

  async lint(text: string, selector: VoiceSelector = {}): Promise<{ ok: boolean; issues: LintIssue[] }> {
    const response = await fetch(new URL("/v1/voice/lint", this.baseUrl), {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        text,
        profile: selector.profile,
        audience: selector.audience,
        surface: selector.surface,
        tone: selector.tone,
      }),
    });
    if (!response.ok) throw new Error(`VoiceMD lint failed: ${response.status} ${await response.text()}`);
    return (await response.json()) as { ok: boolean; issues: LintIssue[] };
  }
}
